overviewer_core/src/primitives/overlay-bounds.c

```diff
@@ -0,0 +1,246 @@
+ * your option) any later version.
+ *
+ * Minecraft Overviewer is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+ * Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with the Overviewer.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "overlay.h"
+
+typedef enum { false, true } bool;
+
+typedef struct {
+    /* inherits from overlay */
+    RenderPrimitiveOverlay parent;
+    void *bounds;
+    int numcolors;
+} RenderPrimitiveBounds;
+
+struct Condition{
+    int minx, minz, maxx, maxz;
+};
+
+
+struct Color {
+    int numconds;
+    struct Condition *conditions;
+    unsigned char r, g, b, a;
+};
+
+/*
+* Determines if point x,z is inside bounds minx, minz, maxx, maxz
+*/
+static int is_inside(int x, int z, int minx, int minz, int maxx, int maxz) {
+	return (x >= minx && x <= maxx && z >= minz && z <= maxz);
+}
+
+static void get_color(void *data,
+                      RenderState *state,
+                      unsigned char *r,
+                      unsigned char *g,
+                      unsigned char *b,
+                      unsigned char *a) {
+    /**
+     * Calculate the color at the current position and store the values to r,g,b,a.
+     **/
+    RenderPrimitiveBounds *self = (RenderPrimitiveBounds *)data;
+    int x = (state->chunkx * 16) + state->x, z = (state->chunkz * 16) + state->z, col, cond;
+    struct Color *bounds = (struct Color *)(self->bounds);
+    struct Condition * c = NULL;
+    bool any = true;
+	
+    /**
+     * Check for every color in the current point is in the given bounds,
+	 * and color appropriately
+     **/
+    // iterate over all the colors
+    for ( col = 0; col < self->numcolors; col++) {
+		any = false;
+        // iterate over all conditions
+        for (cond = 0; cond < bounds[col].numconds; cond++) {
+            c = (struct Condition *)&bounds[col].conditions[cond];
+            // check current point is in the condition
+            if(is_inside(x, z, c->minx, c->minz, c->maxx, c->maxz)) {
+				any = true;
+            }
+        }
+
+		//if current point is in any of the conditions, draw it this color
+		if (any) {
+			// set the color
+			*r = bounds[col].r;
+			*g = bounds[col].g;
+			*b = bounds[col].b;
+			*a = bounds[col].a;
+			return;
+		}      
+    }
+    return;
+}
+
+static int overlay_bounds_start(void *data, RenderState *state, PyObject *support) {
+    /**
+     * Initializing the search for bounds by parsing the arguments and storing them into
+     * appropriate bounds. If no arguments are passed create and use default values.
+     **/
+    PyObject *opt;
+    RenderPrimitiveBounds* self;
+
+    /* first, chain up */
+    int ret = primitive_overlay.start(data, state, support);
+    if (ret != 0)
```

Comment:
Personally I prefer to always wrap conditional blocks in curly brackets, but I'm not sure what the upstream project's styles (and C, generally) dictates as the standard.